### PR TITLE
Run pyupgrade 2.31.0.

### PR DESCRIPTION
**Pull request by bdice**
This PR runs [pyupgrade](https://github.com/asottile/pyupgrade) to modernize some syntax in the Python code base. (I am preparing a "better code" talk for our team that will reference this PR.)

The kinds of modernizing changes that pyupgrade makes are similar in purpose to those made by clang-tidy (#9860), but for Python. These include things like:
- Using literals for empty dicts `{}` and lists `[]`, because they are a bit faster and more readable than `dict()` or `list()`
- Removing extra parentheses or braces wrapping generator expressions if not needed
- Preferring format strings
- Using canonical error names, e.g. `IOError` became an alias for `OSError` in Python 3.3.
- Replacing the "new-style" class definitions like `class A(object):` with the simpler syntax `class A:`
- Removing unnecessary `__future__` imports that were needed for Python 2 compatibility, like `print_function` and `division`
- Using raw strings for regular expressions, to clarify intent and avoid extra escaping

The [pyupgrade README](https://github.com/asottile/pyupgrade) has a long showcase of other rules.

Notes for reviewers:
- pyupgrade has a lot of opinions on how to use type annotations, and I'm not sure if I like all of them (some may have compatibility issues?). I have skipped over the type annotations for now, to set a more reasonable scope for this PR. See commit 54b16b9 (since reverted) for examples. Please let me know what you think about these changes and I may investigate them further in a follow-up PR!
- I have not added this tool to our CI style checks via a pre-commit hook yet. That can be done in a follow-up PR if we agree that we would like these types of changes to be required in the future.
- I used the argument `--py37-plus` to only perform changes that are compatible with Python 3.7+.
- I can revert the changes to the vendored code in `versioneer.py` if desired. (It's pretty outdated anyway, it looks like we're using versioneer 0.18, released on January 1, 2017.)

### CODE REVIEW

**Review by vyasr — APPROVED** [review]
These changes look good to me. I made some comments, but in hindsight I think a significant chunk of them were in the versioneer files which I think we should just revert as you point out in your comment. No need to change vendored files. You should change the pre-commit hook to exclude those files when you do add it. I'm approving now because I suspect that once you remove the versioneer changes I only have 2-3 other comments and they're pretty trivial.

Regarding annotations, I like some of the changes but not all. Removing quoted types where possible is nice, but I don't really like using the pipe operator for `or` everywhere. I'm not sure what best practices are as far as using built-in types for annotations vs using their aliases from the `typing` module, I'd say let's see what some other projects do? But I definitely don't think we need to make these changes here

File: python/cudf/cudf/tests/test_api_types.py [hunk 1]
```diff
@@ -17,9 +17,7 @@
         (int(), False),
         (float(), False),
         (complex(), False),
-        (str(), False),
```
[the remark this answers is not in the document]
Comment (bdice) [reply python/cudf/cudf/tests/test_api_types.py:20]: It converted `str()` to `""` and I manually removed the duplicates.

File: python/cudf/cudf/tests/test_groupby.py [hunk 2]
```diff
@@ -85,8 +85,7 @@ def make_frame(
 
 
 def get_nelem():
-    for elem in [2, 3, 1000]:
-        yield elem
+    yield from [2, 3, 1000]
```
[the remark this answers is not in the document]
Comment (bdice) [reply python/cudf/cudf/tests/test_groupby.py:88]: Good catch. Change applied.

File: python/cudf/cudf/tests/test_dataframe.py [hunk 3]
```diff
@@ -845,7 +845,7 @@ def test_dataframe_to_string_wide(monkeypatch):
     # Test basic
     df = cudf.DataFrame()
     for i in range(100):
-        df["a{}".format(i)] = list(range(3))
+        df[f"a{i}"] = list(range(3))
```
[the remark this answers is not in the document]
Comment (bdice) [reply python/cudf/cudf/tests/test_dataframe.py:848]: I opened #10204 to apply these suggestions and refactor some nearby tests.

File: python/dask_cudf/dask_cudf/io/orc.py [hunk 4]
```diff
@@ -79,7 +79,9 @@ def read_orc(path, columns=None, filters=None, storage_options=None, **kwargs):
         ex = set(columns) - set(schema)
         if ex:
             raise ValueError(
-                "Requested columns (%s) not in schema (%s)" % (ex, set(schema))
+                "Requested columns ({}) not in schema ({})".format(
+                    ex, set(schema)
+                )
```
[the remark this answers is not in the document]
Comment (bdice) [reply python/dask_cudf/dask_cudf/io/orc.py]: Fixed in ad69bb90f971b94c7be378d7a9ffdf00092a3333.

**Comment by bdice** [comment]
> I think a significant chunk of them were in the versioneer files which I think we should just revert as you point out in your comment. [...] You should change the pre-commit hook to exclude those files when you do add it.

Reverted changes to versioneer. Yup, I'll exclude those files if we get to the point of adding a pre-commit hook for pyupgrade.

> Regarding annotations, I like some of the changes but not all. Removing quoted types where possible is nice, but I don't really like using the pipe operator for `or` everywhere. I'm not sure what best practices are as far as using built-in types for annotations vs using their aliases from the `typing` module, I'd say let's see what some other projects do? But I definitely don't think we need to make these changes here.

I agree, with your general sentiment -- pyupgrade's typing changes are the part I'm least sure about. (As an aside, I do agree with pyupgrade that replacing `Union` with `T1 | T2` and `Optional` with `T | None` is just fine, or perhaps slightly better, because it lets you remove `from typing import ...` statements.) The use of annotations-as-strings vs. annotations-as-objects was discussed extensively in light of some recent PEPs and changes to Python 3.10 (see this [issue from pydantic](https://github.com/samuelcolvin/pydantic/issues/2678), for example). I'm generally uninformed about typing at this level of detail, but I think that typing behavior varies significantly among recent Python versions so I am taking an approach of "if it ain't broke... don't fix it." Is having mypy pass our only requirement for typing support?

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py [hunk 5]
```diff
@@ -40,12 +40,12 @@ def test_roundtrip_from_dask(tmpdir, stats):
     tmpdir = str(tmpdir)
     ddf.to_parquet(tmpdir, engine="pyarrow")
     files = sorted(
-        [
+        (
             os.path.join(tmpdir, f)
             for f in os.listdir(tmpdir)
             # TODO: Allow "_metadata" in list after dask#6047
```
[the remark this answers is not in the document]
Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/io/tests/test_parquet.py:46]: Thanks for the ping on this - will open a PR to address this

**Review by charlesbluca — APPROVED** [review]

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (charlesbluca) [reply python/dask_cudf/dask_cudf/io/tests/test_parquet.py:46]: Opened in https://github.com/rapidsai/cudf/pull/10206

**Review by ajschmidt8 — APPROVED** [review]

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply python/dask_cudf/dask_cudf/io/tests/test_parquet.py]: Awesome thanks

File: python/cudf/cudf/tests/test_api_types.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (vyasr) [reply python/cudf/cudf/tests/test_api_types.py:20]: Ah that makes more sense.

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**